File: swift/Sources/SNES/SNES.swift
import SwiftUI

/// High-level SwiftUI host for one Snes9x ROM.
public struct SNES: View {
    @StateObject private var engine: SNESEngine
    private let configuration: SNESConfiguration
    private let controllerConfiguration: SNESControllerConfiguration

    public init(
        configuration: SNESConfiguration,
        controllerConfiguration: SNESControllerConfiguration = .init()
    ) {
        self.configuration = configuration
        self.controllerConfiguration = controllerConfiguration
        _engine = StateObject(wrappedValue: SNESEngine(configuration: configuration))
    }

    public init(
        rom: URL,
        controllerConfiguration: SNESControllerConfiguration = .init()
    ) {
        self.init(
            configuration: SNESConfiguration(romURL: rom),
            controllerConfiguration: controllerConfiguration
        )
    }

    public var body: some View {
        SNESView(
            engine: engine,
            showsControls: configuration.showsTouchControls,
            controllerConfiguration: controllerConfiguration
        )
        .onAppear {
            if configuration.automaticallyStarts { engine.start() }
        }
        .onDisappear { engine.stop() }
    }
}

File: android/snes/src/main/kotlin/snes9x/SNES.kt
package snes9x

import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.compose.LocalLifecycleOwner

@Composable
fun SNES(
    configuration: SNESConfiguration,
    modifier: Modifier = Modifier,
    controllerConfiguration: SNESControllerConfiguration = SNESControllerConfiguration(),
) {
    val context = LocalContext.current
    val engine = remember(configuration) { SNESEngine(context, configuration) }
    val lifecycleOwner = LocalLifecycleOwner.current

    DisposableEffect(engine, lifecycleOwner) {
        val observer =
            LifecycleEventObserver { _, event ->
                when (event) {
                    Lifecycle.Event.ON_RESUME -> engine.resume()
                    Lifecycle.Event.ON_PAUSE -> engine.pause()
                    else -> Unit
                }
            }
        lifecycleOwner.lifecycle.addObserver(observer)
        if (configuration.automaticallyStarts) engine.start()
        onDispose {
            lifecycleOwner.lifecycle.removeObserver(observer)
            engine.close()
        }
    }

    SNESView(
        engine = engine,
        modifier = modifier,
        showsControls = configuration.showsTouchControls,
        controllerConfiguration = controllerConfiguration,
    )
}
